airbyte-integrations/connectors/source-shortio/source_shortio/source.py

```diff
@@ -0,0 +1,266 @@
+        self, 
+        stream_state: Mapping[str, Any] = None, 
+        stream_slice: Mapping[str, Any] = None,
+        next_page_token: Mapping[str, Any] = None
+    ) -> str:
+        # Get all the links 
+        return f"{self.config['domain_id']}/last_clicks"
+
+    def next_page_token(
+        self, 
+        response: requests.Response
+        ) -> Optional[Mapping[str, Any]]:
+        """
+        This function goes through the API responses and ensures that no more requests are left to take place
+        :return str: min(dt) object from the previous API response.
+        """
+        clicks = json.loads(response.text)
+        try:
+            before_dt = sorted(clicks, key=lambda k: k['dt'], reverse=False)[0]['dt']
+            if self.limit > len(clicks):
+                return None
+            else:
+                return before_dt
+        except IndexError:
+            return None
+
+    def get_updated_state(
+        self, 
+        current_stream_state: MutableMapping[str, Any], 
+        latest_record: Mapping[str, Any]
+    ) -> Mapping[str, any]:
+        """
+        Here we keep track of the state between different syncs to ensure that the data fetched is correct.
+        When the object is created, the datetime is taken and records are fetched until that point.
+        Due to varying duration possibilities this allows to a reproducable set of results"
+        """
+        # This method is called once for each record returned from the API to compare the cursor field value in that record with the current state
+        # we then return an updated state object. If this is the first time we run a sync or no state was passed, current_stream_state will be None.
+        if current_stream_state is not None and 'dt' in current_stream_state:
+            return {'dt' : self.before_dt}
+        else:
+            return {'dt' : self.config['start_date']}
+        
+    def request_body_json(
+        self, 
+        stream_state: Mapping[str, Any], 
+        stream_slice: Mapping[str, any] = None, 
+        next_page_token: Mapping[str, Any] = None
+    ) -> MutableMapping[str, Any]:
+        """
+        This method passes the arguments necessary to get the clicks from the shortio API. 
+        No parameters have been implemented here at all with the exception of hardcoding human clicks only to come through.
+        Human clicks are hardcoded to reduce unnecessary clicks from coming through.
+
+        :return dict: json body for the request
+        """
+        payload =  {
+            "limit" : self.limit,
+            "include" : {"human":True},
+            "beforeDate" :  next_page_token or self.before_dt,
+        }
+
+        if stream_state and 'dt' in stream_state.keys():
+            payload['afterDate'] = stream_state['dt']
+        else:
+            payload['afterDate'] = self.config['start_date']
+
+        # Capturing request
+        self.logger.log(level='INFO', message=str(payload))
```

Comment:
was this for debugging? 

we should remove this, otherwise it will incur significant performance penalty